### ISSUE TITLE
Improve producer operator interoperability with a `ExpressibleBySignalProducer` protocol.

### ISSUE THREAD

**Issue by andersio** [ReactiveCocoa/ReactiveSwift#349]
Introduce a `ExpressibleBySignalProducer` protocol, and constrain the argument of suitable binary and n-ary operators to it. This is a source compatible binary incompatible change.

`Signal`, `SignalProducer` and `PropertyProtocol` would all conform to `SignalProducerRepresentable`, and returns its representation as a `SignalProducer` as requested. This allows all producer operators to accept any reactive primitives without defining explicit overloads.

e.g.
```swift
extension SignalProducer {
    public func combineLatest<Other>(with: Other) -> SignalProducer<(Value, Other.Value), Error>
    where Other: ExpressibleBySignalProducer, Other.Error == Error
} 

// All valid.
producer.combineLatest(with: signal)
producer.combineLatest(with: producer)
producer.combineLatest(with: property)

// Mix and match.
SignalProducer.combineLatest(signal, producer, property)
SignalProducer.combineLatest(property, producer, signal)
```

Additionally, new overloads may be added to improve the `NoError` usability.
```swift
extension SignalProducer {
    public func combineLatest<Other>(with: Other) -> SignalProducer<(Value, Other.Value), Error>
    where Other: SignalProducerConvertible, Other.Error == NoError
}

extension SignalProducer where Error == NoError {
    public func combineLatest<Other>(with: Other) -> SignalProducer<(Value, Other.Value), Other.Error>
    where Other: SignalProducerConvertible
}
```

This may also reduce the number of flattening operator overloads, and may allow the removal of `SignalProtocol` and `SignalProducerProtocol`.

### Note
`ExpressibleBySignal` and `ExpressibleByProperty` were considered, but:

1. Property is the most restricting contract in ReactiveSwift, so practically the only `ExpressibleByProperty` entities are properties, rendering the protocol useless.

1. `ExpressibleBySignal` can ideally cover both producers and properties. But doing so is equivalent to lossy type narrowing, since the events emitted during the producer starting side effect would be lost.

**Comment by mdiep** [comment]
A small PR that demonstrates this with a single operator would probably helpful in evaluating this proposal.

**Comment by sharplet** [comment]
Left a comment on https://github.com/ReactiveCocoa/ReactiveSwift/pull/351#issuecomment-304308268 about naming.

**Comment by andersio** [comment]
Implemented in #351.
